Add Menu `Calibration -> Temp. model cal`

3d-gussner opened the pull request:
Runs M310 A F0
Enables with M310 S1
Saves settings M500

Tested on MK404 with Octoprint serial logging.

[this build](https://app.travis-ci.com/github/prusa3d/Prusa-Firmware/jobs/586714582#L625) - [previous build](https://app.travis-ci.com/github/prusa3d/Prusa-Firmware/jobs/586414833#L625)
Flash: 257168 - 256965 = +203
RAM: 5396 - 5395 = +1


ToDos:

- [X] cherry-pick to MK3_3.12

3d-gussner committed:
Temp Model cal. menu

3d-gussner committed:
Fix MK25/S builds

DRracer commented on Firmware/temperature.cpp:
```diff
@@ -2777,21 +2787,24 @@ static bool autotune(int16_t cal_temp)
     for(uint8_t i = 0; i != 2; ++i) {
-        const char* PROGMEM verb = (i == 0? PSTR("initial"): PSTR("refining"));
-
+        const char* PROGMEM verb = (i == 0? PSTR("inital"): PSTR("refine"));
```
typo: inital -> initial

3d-gussner replied:
fixed

DRracer commented on Firmware/temperature.cpp:
```diff
@@ -2871,30 +2888,33 @@ static bool autotune(int16_t cal_temp)
-        SERIAL_ECHOLNPGM("TM: printer needs to be idle for calibration");
+        sprintf_P(tm_message, PSTR("TM: Cal. NOT IDLE"));
+        SERIAL_ECHOLN(tm_message);
+        lcd_setstatuspgm(tm_message);
```
`lcd_setstatuspgm` with a RAM resident variable? Shouldn't it have been just `lcd_setstatus`?

DRracer commented on Firmware/temperature.cpp:
```diff
@@ -2871,30 +2888,33 @@ static bool autotune(int16_t cal_temp)
-        lcd_setstatuspgm(_i("TM autotune failed")); ////MSG_TM_AUTOTUNE_FAILED c=20
+        sprintf_P(tm_message, PSTR("TM: calibr. failed!"));
+        SERIAL_ECHOLN(tm_message);
+        lcd_setstatuspgm(tm_message); ////MSG_TM_AUTOTUNE_FAILED c=20
```
`lcd_setstatuspgm` again, see above.

3d-gussner replied:
fixed

DRracer commented on Firmware/temperature.cpp:
```diff
@@ -2811,7 +2824,9 @@ static bool autotune(int16_t cal_temp)
         wait(30000); // settle PID regulation
 
-        printf_P(PSTR("TM: %S R estimation @ %dC\n"), verb, cal_temp);
+        sprintf_P(tm_message, PSTR("TM: %S R %dC"), verb, cal_temp);
```
since these 3-line sequences are repeated multiple times in the code, I'd suggest making a common function for them:
```
sprintf_P(tm_message, something, ...);
SERIAL_ECHOLN(tm_message);
lcd_setstatus(tm_message);
```
It may save some code...

3d-gussner replied:
saved 2 bytes by adding lcd_setstatus_serial

DRracer reviewed (changes requested):
please see my comments in the code

3d-gussner committed:
Fix typo
Add lcd_setatatus_serial to output lcd status message and serial at the same time

wavexx commented on Firmware/temp_model.h:
```diff
@@ -59,6 +59,7 @@ struct model_data
 };
 
 static bool enabled;          // model check enabled
+static bool valid = false;    // model is valid
```
I'm not sure why this is needed, can you explain?
temp_model::calibrated() will return true if the data is ok, so we shouldn't need this extra variable?

DRracer reviewed (approved):
approving, the decision about the `valid` flag will be done later
